### DISCUSSION TITLE
Option in <Link> component to maintain certain search params

### DISCUSSION THREAD

**Discussion by brightpixels** [remix-run/remix#3647] [General]
Hello,
When we toggle language, we do it via the url as `?lng=en` (using remix-i18next lib). However, the lang search params disappears from the url when a link on the page is clicked on that has it's own search params (like pagination) which means language is reset back to the default (this also happens on form submission that redirect to a new page if successful or reloads with errors). I think this is intentional by Remix? To get around this, we are having to useSearchParams and if there is a lng present, then to append it ourselves manually. Doing this for all links is a big overhead. I was wondering if there is any plans to make this type of thing easier?

**Comment by sergiodxa** [comment]
You should not use the search params as a persistent way to keep the user preferred locale, use a cookie or the session instead.

**Reply by brightpixels** [reply]
Hello. We're using the lng search parameter to change the language (the app must work without client JavaScript). And then persist the current detected language in the cookie. So user language selection is never lost (my earlier statement wasn’t accurate).

The problem is when the user is clicking on pagination links, they all get lost when user switch language which reloads the current page URL with the new language search params and removes existing pagination search params, effectively resetting the user back to page 1 in the pagination.

**Comment by kiliman** [comment]
I have a helper function `useScopedParams` that make it easy to maintain search params across links and navigations.

demo: https://remix-scoped-params-production.up.railway.app
